benchmarks/humaneval.py

```diff
@@ -0,0 +1,205 @@
+        accuracy = correct / total
+        ret_dict[key]["correct"] = correct
+        ret_dict[key]["total"] = total
+        ret_dict[key]["accuracy"] = accuracy
+
+    return ret_dict
+
+
+def pass_at_k(k: int, stat: str) -> Metric:
+    def _metric(correct: int, total: int) -> float:
+        if total - correct < k:
+            return 1.0
+        return 1.0 - np.prod(1.0 - k / np.arange(total - correct + 1, total + 1))
+
+    def batch_aggregate_metric(scores: list[Score]) -> float:
+        id_to_stats = _per_sample_collator(scores)
+        list_of_metrics = []
+        for id in id_to_stats:
+            n, c = id_to_stats[id]["total"], id_to_stats[id]["correct"]
+            list_of_metrics.append(_metric(c, n))
+        if stat == "mean":
+            return np.mean(list_of_metrics).item()
+        else:
+            return np.std(list_of_metrics).item()
+
+    return batch_aggregate_metric
+
+
+@metric("accuracy")
+def sample_aggregate_accuracy() -> Metric:
+    def metric(scores: list[Score]) -> float:
+        id_to_stats = _per_sample_collator(scores)
+        list_of_metrics = [id_to_stats[k]["accuracy"] for k in id_to_stats]
+        return np.mean(list_of_metrics).item()
+
+    return metric
+
+
+@metric("pass@1:mean")
+def pass_at_1_mean() -> Metric:
+    return pass_at_k(1, "mean")
+
+
+@metric("pass@2:mean")
+def pass_at_2_mean() -> Metric:
+    return pass_at_k(2, "mean")
+
+
+@metric("pass@5:mean")
+def pass_at_5_mean() -> Metric:
+    return pass_at_k(5, "mean")
+
+
+@metric("pass@1:std")
+def pass_at_1_std() -> Metric:
+    return pass_at_k(1, "std")
+
+
+@metric("pass@2:std")
+def pass_at_2_std() -> Metric:
+    return pass_at_k(2, "std")
+
+
+@metric("pass@5:std")
+def pass_at_5_std() -> Metric:
+    return pass_at_k(5, "std")
+
+
+@solver
+def find_code():
+    async def solve(state: TaskState, generate: Generate) -> TaskState:
+        pattern = re.compile(r"```python\n(.*?)```", re.DOTALL)
+        matches = pattern.findall(state.output.completion)
+        extracted_answer = matches[0] if len(matches) >= 1 else state.output.completion
+        extracted_answer = extracted_answer[
+            extracted_answer.find(":\n    ") + 2 :
+        ]  # remove signature
+        state.output.completion = extracted_answer
+        return state
+
+    return solve
+
+
+@scorer(
+    metrics=[
+        sample_aggregate_accuracy(),
+        pass_at_1_mean(),
+        pass_at_1_std(),
+        pass_at_2_mean(),
+        pass_at_2_std(),
+        pass_at_5_mean(),
+        pass_at_5_std(),
+    ]
+)
+def verify_correctness() -> Scorer:
+    async def score(state: TaskState, target: Target) -> Score:
+        code = [
+            state.metadata["prompt"],
+            state.output.completion,
+            "\n",
+            state.metadata["test"],
+            "\n",
+            "".join(["check(", state.metadata["entry_point"], ")"]),
+        ]
+
+        result = await subprocess(
```

Comment:
Here I think you will want to use `tool_environment().exec()` rather than subprocess (it has all the same arguments). You would also want to add `tool_environment="local"` to the task definition.

This will have the benefit of being able to do this at the user's option:

```bash
inspect eval humaneval.py --toolenv docker
```

And the scoring would in that case run in a sandboxed Docker container (or any other configured tool environment, see https://inspect.ai-safety-institute.org.uk/agents.html#sec-tool-environments).

Reply:
Thanks for the feedback JJ. I initially thought of using tool environments but had a misunderstanding that they should be used during model generation explicitly (as opposed to during scoring, which is the case here). 

I think it’s a great point regarding containerisation so I’ll make the relevant change, validate it, and push it back here for review.